### PR TITLE
feat: support external collection schema evolution

### PR DESCRIPTION
**Pull request by weiliu1031**
issue: #45881

design doc: docs/design-docs/design_docs/20260605-external-table-schema-evolution.md

### What changed

This PR supports schema evolution for external collections.

It allows `AlterCollectionSchema` to update external collection schemas,
validates source-backed fields and function-output fields, adds SDK helpers,
and records refresh schema versions.

Refresh jobs recheck the captured schema version against DataCoord's current
schema cache before committing. Once DataCoord observes a newer schema, stale
refresh results are rejected. Cross-component serialization for the narrower
RootCoord-to-DataCoord cache propagation window remains out of scope.

It also ensures schema-visible external fields are not served until their
backing manifests are materialized. Query/search/index paths reject
unmaterialized fields, QueryCoord schedules loaded segment reopen when
manifests are stale, and QueryNode refreshes BM25/function state after schema
changes.

External collection schema-mutation WAL messages are unreplicable, including
the legacy add-field compatibility path. CDC therefore does not apply an
external schema change without the refresh and manifest updates needed to
materialize it. Internal collection schema evolution remains replicable.

### Validation

- `git diff --check`
- Targeted RootCoord tests for external and internal replication flags
- Targeted Go SDK external add-column refresh test

### CODE REVIEW

File: internal/core/src/segcore/ChunkedSegmentSealedImpl.cpp [hunk 1]
```diff
@@ -733,7 +749,14 @@ ChunkedSegmentSealedImpl::LoadColumnGroups(const std::string& manifest_path,
         std::vector<FieldId> field_ids;
         field_ids.reserve(cg->columns.size());
         for (auto& column : cg->columns) {
-            field_ids.emplace_back(schema_->ResolveColumnFieldId(column));
+            auto field_id = resolve_loadable_column_field(schema_, column);
```
Comment (liliu-z) [inline internal/core/src/segcore/ChunkedSegmentSealedImpl.cpp]: LoadColumnGroups silently continues over any manifest column that resolve_loadable_column_field cannot map to a current-schema field. This path used to throw DataFormatBroken before the PR changed ResolveColumnFieldId to return std::nullopt, so genuine manifest corruption now drops columns with no log or metric, indistinguishable from benign schema-evolution drift (a dropped field). Add a scoped log on skip so operators retain a signal for real inconsistency, while avoiding a per-column WARN that would spam on every dropped-field load.

File: internal/datanode/external/task_update.go [hunk 2]
```diff
@@ -19,10 +19,11 @@ package external
 // RefreshExternalCollectionTask handles updating external collection segments by fetching fragments from external sources
 // and organizing them into segments with balanced row counts.
 //
-// SEGMENT ID ALLOCATION WORKFLOW:
-// - DataCoord pre-allocates a batch of segment IDs (default 1000) via allocator.AllocN()
+// ID ALLOCATION WORKFLOW:
+// - DataCoord pre-allocates a batch of IDs (default 10000) via allocator.AllocN()
```
Comment (liliu-z) [inline internal/datanode/external/task_update.go]: The doc comment says DataCoord pre-allocates a batch of IDs (default 10000), but the governing param ExternalCollectionPreAllocSegments — consumed at task_refresh_external_collection.go:672-674 for AllocN — defaults to 500000. Update the comment to 500000. Note the reviewer cited internal/rootcoord/task_update.go, which does not exist; the comment lives in internal/datanode/external/task_update.go:23.

File: internal/core/src/segcore/SegmentGrowingImpl.cpp [hunk 3]
```diff
@@ -2766,10 +2766,19 @@ SegmentGrowingImpl::LoadColumnsGroups(std::string manifest_path) {
     auto schema_snapshot = schema_;
     auto group_has_live_field = [&](int64_t group_index) {
         for (const auto& column : column_groups->at(group_index)->columns) {
-            // An untranslatable column name is a broken manifest and throws
-            // (same contract as the sealed loader).
+            // An untranslatable column name is a broken manifest and throws.
             auto field_id = schema_snapshot->ResolveColumnFieldId(column);
-            if (schema_snapshot->has_field(field_id)) {
+            if (!field_id.has_value()) {
+                if (schema_snapshot->is_external_collection()) {
```
Comment (liliu-z) [inline internal/core/src/segcore/SegmentGrowingImpl.cpp:3183]: The growing-segment loader and the sealed-segment loader diverge on how they treat an external manifest column that no longer resolves to a live field after a drop. SegmentGrowingImpl::LoadColumnsGroups throws DataFormatBroken when ResolveColumnFieldId returns nullopt for an external collection, while ChunkedSegmentSealedImpl treats the same case as a legal drop leftover and skips it. This contradicts the PR's own drop-tolerance intent (both sites comment that dropped columns are legal leftovers). For milvus-table collections it is not triggered — manifest columns are numeric, so ResolveColumnFieldId's numeric fallback (Schema.cpp:504-506) returns a value and only has_field()==false skips the group — but a source-named (non-numeric) external column whose mapped field was dropped resolves to nullopt and would fail a growing-segment load that the sealed loader handles gracefully. Make the growing loader skip nullopt columns like the sealed loader.

Comment (weiliu1031) [reply internal/core/src/segcore/SegmentGrowingImpl.cpp:3183]: This case is not reachable for external collections. Non-milvus-table manifests can contain source-named columns, but external collections are read-only and do not produce growing segments. Segments created by `RefreshExternalCollection` are normalized to `SegmentState_Flushed` and loaded through the sealed-segment path. `SegmentGrowingImpl` is only used for segments from `UnflushedSegmentIds`, which contains growing segments. Therefore, a source-named external column cannot reach `SegmentGrowingImpl::LoadColumnsGroups` in the supported external-collection lifecycle. The sealed-loader behavior is the relevant path for drop tolerance; changing the growing loader would only handle an invalid state, not a reachable issue in this PR.

File: internal/rootcoord/ddl_callbacks_alter_collection_schema.go [hunk 4]
```diff
@@ -118,11 +119,24 @@ func (c *Core) broadcastAlterCollectionSchemaAdd(ctx context.Context, broadcaste
...
 	if err != nil {
 		return err
 	}
-	if err := validateSchemaEvolution(coll, schema); err != nil {
+	schema, properties, err := buildAlterSchemaAddSchema(coll, plan, assignedFieldID)
+	if err != nil {
+		return err
+	}
+	if sourceSchema != nil {
+		if err := typeutil.ValidateMilvusTableSchemaIdentity(schema, sourceSchema, true); err != nil {
```
Comment (liliu-z) [inline internal/rootcoord/ddl_callbacks_alter_collection_schema.go:132]: This makes add-field impossible for a normal non-nullable milvus-table source field. Proxy requires every added ordinary field to be nullable, but this identity check includes Nullable. false is rejected in Proxy; true is rejected here. Please allow source non-nullable -> target nullable, or change the contract and cover it with a test.

File: internal/rootcoord/ddl_callbacks_alter_collection_schema.go [hunk 5]
```diff
@@ -118,11 +119,24 @@ func (c *Core) broadcastAlterCollectionSchemaAdd(ctx context.Context, broadcaste
...
+			return merr.Wrap(err, "milvus-table target schema does not match source snapshot schema")
+		}
+	}
+	var allowedHistoricalFieldIDs []int64
+	if sourceSchema != nil && plan.Kind == schemautil.AlterSchemaAddField {
+		allowedHistoricalFieldIDs = []int64{assignedFieldID}
```
Comment (liliu-z) [inline internal/rootcoord/ddl_callbacks_alter_collection_schema.go:138]: This exception can reuse a dropped target-only ID. Example: function output 106 is materialized and dropped; the source later adds field 106; old manifest/ChildFields 106 can then make the new source field look already materialized and may serve the old output. We need an ID namespace/tombstone here, or reject fields outside the creation-time source snapshot.

File: internal/datacoord/task_refresh_external_collection.go [hunk 6]
```diff
@@ -351,11 +352,18 @@ func applyExternalCollectionSegmentUpdate(
...
-	// Build update operators
 	var operators []UpdateOperator
 	var patchErr error
 
 	validationOperator := func(modPack *updateSegmentPack) bool {
+		collection := modPack.meta.GetCollection(collectionID)
+		if collection == nil || collection.Schema == nil {
+			return modPack.fail(merr.WrapErrServiceInternalMsg("collection %d not found in meta", collectionID))
+		}
+		if collection.Schema.GetVersion() != expectedSchemaVersion {
```
Comment (liliu-z) [inline internal/datacoord/task_refresh_external_collection.go:360]: This only checks DataCoord's local schema cache. Alter commits V2 in RootCoord before pushing V2 to DataCoord, so an old refresh can take segMu in that gap, still see V1 here, commit, and finish successfully. That breaks the stale-refresh fence described by the design. Please make the final commit serialize/CAS against the authoritative schema update and add a deterministic race test.

File: tests/go_client/testcases/external_table_function_test.go [hunk 7]
```diff
@@ -1065,3 +1267,547 @@ teiRefreshDone:
...
+		WithOutputFields("id"))
+	common.CheckErr(t, err, true)
+
+	requireExternalQueryFieldMissing(t, ctx, mc, collName, "category")
+
+	refreshResult, err = mc.RefreshExternalCollection(ctx, client.NewRefreshExternalCollectionOption(collName))
+	common.CheckErr(t, err, true)
+	waitRefreshTerminal(t, ctx, mc, refreshResult.JobID, entity.RefreshStateCompleted)
+
+	err = mc.ReleaseCollection(ctx, client.NewReleaseCollectionOption(collName))
```
Comment (liliu-z) [inline tests/go_client/testcases/external_table_function_test.go]: Releasing here bypasses the online Reopen path this feature depends on. Refresh completion only proves DataCoord materialization. Please keep the collection loaded and wait until category is queryable, otherwise the main serving transition is still untested.

File: tests/go_client/testcases/external_table_function_test.go [hunk 8]
```diff
@@ -1065,3 +1267,547 @@ teiRefreshDone:
...
+		client.NewAlterCollectionSchemaAddFunctionOption(collName, bm25Fn, bm25Field))
+	common.CheckErr(t, err, true)
+
+	common.CheckErr(t, searchDense(), true)
+
+	refreshResult, err = mc.RefreshExternalCollection(ctx, client.NewRefreshExternalCollectionOption(collName))
+	common.CheckErr(t, err, true)
+	waitRefreshTerminal(t, ctx, mc, refreshResult.JobID, entity.RefreshStateCompleted)
+
+	idxTask, err = mc.CreateIndex(ctx, client.NewCreateIndexOption(collName, "bm25_sparse", sparseIdx))
```
Comment (liliu-z) [inline tests/go_client/testcases/external_table_function_test.go]: bm25_sparse already has bound IndexParams. This explicit CreateIndex after refresh can retrigger the index and hide a broken bound-index inspector path. Please verify no build before coverage, automatic build after refresh, and online Reopen/search without this call.

File: docs/design-docs/design_docs/20260605-external-table-schema-evolution.md [hunk 9]
```diff
@@ -0,0 +1,936 @@
...
+
+| Request shape | Operation |
+|---------------|-----------|
+| `AddRequest.field_infos` has exactly one field and `func_schema` is empty | Add source-backed field |
+| `AddRequest.field_infos` contains exactly one new output field and `func_schema` has exactly one function | Add function or add BM25 |
+| `DropRequest.field_name` or `DropRequest.field_id` | Drop source-backed field |
+| `DropRequest.function_name` | Drop function or drop BM25 |
+
+`do_physical_backfill=true` is rejected for external collections. Physical
+materialization is driven by `RefreshExternalCollection`. For internal
```
Comment (liliu-z) [inline docs/design-docs/design_docs/20260605-external-table-schema-evolution.md:198]: Need an explicit replication contract here. AlterCollection schema DDL is replicable, while RefreshExternalCollection and BatchUpdateManifest are WithUnreplicable. A replica can otherwise get the new schema/index metadata without the materialization. Either reject this combination under CDC or replicate the materialization protocol.

File: tests/go_client/testcases/external_table_function_test.go [hunk 10]
```diff
@@ -1065,3 +1267,547 @@ teiRefreshDone:
...
+	common.CheckErr(t, err, true)
+	common.CheckErr(t, idxTask.Await(ctx), true)
+
+	loadTask, err := mc.LoadCollection(ctx, client.NewLoadCollectionOption(collName))
+	common.CheckErr(t, err, true)
+	common.CheckErr(t, loadTask.Await(ctx), true)
+
+	categoryField := entity.NewField().WithName("category").WithDataType(entity.FieldTypeVarChar).
+		WithMaxLength(1024).WithExternalField("text_zh").WithNullable(true)
+	err = mc.AlterCollectionSchema(ctx, client.NewAlterCollectionSchemaAddFieldOption(collName, categoryField))
```
Comment (liliu-z) [inline tests/go_client/testcases/external_table_function_test.go:1686]: This path is missing the DataCoord cold-cache case. BroadcastAlteredCollection currently initializes partitionIDs with make(len) and then appends, so a one-partition collection is cached as [0, realID] and the next refresh fails. Please fix the slice and add an alter-before-first-refresh test.

File: internal/rootcoord/ddl_callbacks_alter_collection_schema.go (hunk 4 above)
Comment (weiliu1031) [reply internal/rootcoord/ddl_callbacks_alter_collection_schema.go:132]: Addressed in the current branch update. The milvus-table identity check now permits widening a non-nullable source field to a nullable target field, while still rejecting the reverse nullable-to-non-nullable narrowing. Unit coverage was added for both directions.

File: internal/rootcoord/ddl_callbacks_alter_collection_schema.go (hunk 5 above)
Comment (weiliu1031) [reply internal/rootcoord/ddl_callbacks_alter_collection_schema.go:138]: Addressed in the current branch update. We now persist a monotonically growing `milvus_table_target_only_field_ids` reservation in the collection schema properties. Virtual PK and generated-output IDs remain permanently reserved after drop, and a later source snapshot that reuses one of those IDs is rejected. Add/drop and legacy-metadata cases are covered by tests.

File: internal/datacoord/task_refresh_external_collection.go (hunk 6 above)
Comment (weiliu1031) [reply internal/datacoord/task_refresh_external_collection.go:360]: Thanks for calling this out. The final refresh commit already rechecks the durable job schema version against the DataCoord collection schema while applying the segment changes atomically, so the normal concurrent-alter case is rejected once DataCoord observes the new schema. We agree that this does not close the narrower RootCoord-committed/DataCoord-cache-not-yet-updated window described here. We are intentionally not adding cross-component serialization or CAS in this PR and will keep this as a known corner case for follow-up. A one-shot RootCoord schema RPC would not fully solve it because a check-to-commit TOCTOU window would still remain.

File: tests/go_client/testcases/external_table_function_test.go (hunk 7 above)
Comment (weiliu1031) [reply tests/go_client/testcases/external_table_function_test.go]: Addressed in the current branch update. The loaded add-field E2E now keeps the collection loaded across refresh and waits until `category` is queryable, so it exercises the online Reopen path instead of release/reload. The segcore reopen path also treats a changed external manifest as replacement loading, allowing unchanged fields from the new manifest to replace already-loaded data instead of triggering the duplicate-field assertion.

File: tests/go_client/testcases/external_table_function_test.go (hunk 8 above)
Comment (weiliu1031) [reply tests/go_client/testcases/external_table_function_test.go]: Addressed in the current branch update. The explicit `CreateIndex` for `bm25_sparse` was removed. The test now verifies that the bound index metadata exists but has zero indexed rows before refresh, verifies that no build starts before field coverage is materialized, waits for the automatic build after refresh, and then searches while the collection remains loaded to cover online Reopen.

File: docs/design-docs/design_docs/20260605-external-table-schema-evolution.md (hunk 9 above)
Comment (weiliu1031) [reply docs/design-docs/design_docs/20260605-external-table-schema-evolution.md:198]: Addressed in the pending CDC update. External-collection schema-mutation messages are marked unreplicable for `AlterCollectionSchema` add/drop and the legacy `AddCollectionField` compatibility path, while internal collection schema evolution remains replicable. The design document now states this CDC contract, and unit tests cover both external and internal message behavior.

File: tests/go_client/testcases/external_table_function_test.go (hunk 10 above)
Comment (weiliu1031) [reply tests/go_client/testcases/external_table_function_test.go:1686]: Addressed in the current branch update. `BroadcastAlteredCollection` now builds partition IDs with zero length and preallocated capacity, removing the zero-value prefix. DataCoord also refreshes the cached partition IDs from the broadcast. A broker unit test and an alter-before-first-refresh E2E cover the cold-cache path.
